include/revng/Pipeline/Target.h

```diff
@@ -235,21 +212,73 @@ class TargetsList {
   template<typename... Args>
   void emplace_back(Args &&...A) {
     Contained.emplace_back(std::forward<Args>(A)...);
-    removeDuplicates();
+    llvm::sort(Contained);
+    Contained.erase(unique(Contained.begin(), Contained.end()),
+                    Contained.end());
   }
 
   void merge(const TargetsList &Other);
 
   void push_back(const Target &Target) {
     Contained.push_back(Target);
-    removeDuplicates();
+    llvm::sort(Contained);
+    Contained.erase(unique(Contained.begin(), Contained.end()),
+                    Contained.end());
   }
 
   template<typename... Args>
   auto erase(Args &&...A) {
     return Contained.erase(std::forward<Args>(A)...);
   }
 
+private:
+  struct Comp {
+    bool operator()(const Target &T, const Kind &K) const {
+      return &T.getKind() < &K;
+    }
+    bool operator()(const Kind &K, const Target &T) const {
+      return &K < &T.getKind();
+    }
+  };
+
+public:
+  TargetsList extracEntriesOfKind(const Kind &K) {
+    auto [b, e] = std::equal_range(begin(), end(), K, Comp());
+    TargetsList ToReturn;
+    copy(b, e, std::back_inserter(ToReturn));
+    Contained.erase(b, e);
+    return ToReturn;
+  }
+
+  TargetsList copyEntriesOfKind(const Kind &K) const {
+    auto [b, e] = std::equal_range(begin(), end(), K, Comp());
+    TargetsList ToReturn;
+    copy(b, e, std::back_inserter(ToReturn));
+    return ToReturn;
+  }
```

Comment:
I feel like these are too coupled with its users.
I'd consider simply having `removeByKind` and let users do `llvm::copy(entriesOfKind())`.